Replace enforce_dtype with dtype in ShardedTensor.gather

yingufan opened the pull request:
Summary:
Sometimes local_shards are empty on some ranks, and out.dtype is float16, which will cause error if enforce_dtype is True because `data` will be float32.

Callers know best what dtype they want, so we can just let callers decide.

Temporarily keep enforce_dtype for backward compatibility

Test Plan: Run local and MAST job

Reviewed By: uciyc123

Differential Revision: D46886551

wanchaol commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -387,7 +388,9 @@ def gather(  # type: ignore[override]
                 Default: ``None``
-            enforce_dtype: force the imterediate tensor with the same type as input and output
+            enforce_dtype: Deprecated.
+            dtype: force intermediate tensors to have this dtype.
```
nit: force the gathered tensor to be this dtype

wanchaol reviewed (approved):
lgtm

wanchaol commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -387,7 +388,9 @@ def gather(  # type: ignore[override]
                 Must to be provided ONLY on ``dst`` rank.
                 Default: ``None``
-            enforce_dtype: force the imterediate tensor with the same type as input and output
+            enforce_dtype: Deprecated.
```
Copying over the comment: can you still keep the enforce_dtype logic `if dtype is None and enfore_dtype is True`? This help us maintain BC, you can throw some warnings in this case to warn user that enforce_dtype is being deprecated

yingufan committed:
Replace enforce_dtype with dtype in ShardedTensor.gather (#110561)

Summary:

Sometimes local_shards are empty on some ranks, and out.dtype is float16, which will cause error if enforce_dtype is True because `data` will be float32
https://fburl.com/conveyor/ohcbsmqq

Callers know best what dtype they want, so we can just let callers decide.

Temporarily keep enforce_dtype for backward compatibility

Test Plan: Run local and MAST job

Reviewed By: uciyc123

Differential Revision: D46886551
